ESPHome fails to build with esp-idf framework if wired ethernet is used

### The problem

Build fails


### Which version of ESPHome has the issue?

2023.6.2

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2023.6.3

### What platform are you using?

ESP32-IDF

### Board

wt32-eth01

### Component causing the issue

components/ethernet

### Example YAML snippet

```yaml
`
esp32:
  board: wt32-eth01
  framework:
    type: esp-idf
    version: 5.0.2
    platform_version: 6.3.2
    # Custom sdkconfig options
    sdkconfig_options:
      CONFIG_COMPILER_OPTIMIZATION_SPEED: y
      CONFIG_GPIO_CTRL_FUNC_IN_IRAM: y
`
`
ethernet:
  type: LAN8720
  mdc_pin: GPIO23
  mdio_pin: GPIO18
  clk_mode: GPIO0_IN
  phy_addr: 1
  power_pin: GPIO16
`
```


### Anything in the logs that might be useful for us?

```txt
`
INFO ESPHome 2023.6.2
INFO Reading configuration /config/esphome/climate.yaml...
WARNING The selected ESP-IDF framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
WARNING The selected ESP-IDF framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
INFO Generating C++ source...
INFO Updating https://github.com/espressif/esp-protocols.git@mdns-v1.0.9
INFO Compiling app...
Processing climate (board: wt32-eth01; framework: espidf; platform: platformio/espressif32@6.3.2)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.50002.230601 (5.0.2) 
 - tool-cmake @ 3.16.4 
 - tool-ninja @ 1.7.1 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 11.2.0+2022r1
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.4
Compiling /data/climate/.pioenvs/climate/src/esphome/components/ethernet/esp_eth_phy_jl1101.o
src/esphome/components/ethernet/esp_eth_phy_jl1101.c:22:10: fatal error: eth_phy_regs_struct.h: No such file or directory

*****************************************************************************
* Looking for eth_phy_regs_struct.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:eth_phy_regs_struct.h"
* Web  > https://registry.platformio.org/search?q=header:eth_phy_regs_struct.h
*
*****************************************************************************

   22 | #include "eth_phy_regs_struct.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
*** [/data/climate/.pioenvs/climate/src/esphome/components/ethernet/esp_eth_phy_jl1101.o] Error 1
========================== [FAILED] Took 4.64 seconds ==========================
`
```


### Additional information

This happens because components\ethernet/esp_eth_phy_jl1101.c unconditionally includes eth_phy_regs_struct.h, which is part of Arduino framework (which is not available when building with esp-idf framework)

--- a/esphome/components/ethernet/esp_eth_phy_jl1101.c
+++ b/esphome/components/ethernet/esp_eth_phy_jl1101.c
@@ -19,7 +19,11 @@
 #include <sys/cdefs.h>
 #include "esp_log.h"
 #include "esp_eth.h"
+#if ESP_IDF_VERSION_MAJOR >= 5
+#include "esp_eth_phy_802_3.h"
+#else
 #include "eth_phy_regs_struct.h"
+#endif
 #include "freertos/FreeRTOS.h"
 #include "freertos/task.h"
 #include "driver/gpio.h"
@@ -170,7 +174,11 @@ static esp_err_t jl1101_reset_hw(esp_eth_phy_t *phy) {
   return ESP_OK;
 }
 
+#if ESP_IDF_VERSION_MAJOR >= 5
+static esp_err_t jl1101_negotiate(esp_eth_phy_t *phy, eth_phy_autoneg_cmd_t cmd, bool *nego_state) {
+#else
 static esp_err_t jl1101_negotiate(esp_eth_phy_t *phy) {
+#endif
   phy_jl1101_t *jl1101 = __containerof(phy, phy_jl1101_t, parent);
   esp_eth_mediator_t *eth = jl1101->eth;
   /* in case any link status has changed, let's assume we're in link down status */
@@ -285,7 +293,11 @@ static esp_err_t jl1101_init(esp_eth_phy_t *phy) {
   esp_eth_mediator_t *eth = jl1101->eth;
   // Detect PHY address
   if (jl1101->addr == ESP_ETH_PHY_ADDR_AUTO) {
+#if ESP_IDF_VERSION_MAJOR >= 5
+    PHY_CHECK(esp_eth_phy_802_3_detect_phy_addr(eth, &jl1101->addr) == ESP_OK, "Detect PHY address failed", err);
+#else
     PHY_CHECK(esp_eth_detect_phy_addr(eth, &jl1101->addr) == ESP_OK, "Detect PHY address failed", err);
+#endif
   }
   /* Power on Ethernet PHY */
   PHY_CHECK(jl1101_pwrctl(phy, true) == ESP_OK, "power control failed", err);
@@ -324,7 +336,11 @@ esp_eth_phy_t *esp_eth_phy_new_jl1101(const eth_phy_config_t *config) {
   jl1101->parent.init = jl1101_init;
   jl1101->parent.deinit = jl1101_deinit;
   jl1101->parent.set_mediator = jl1101_set_mediator;
+#if ESP_IDF_VERSION_MAJOR >= 5
+  jl1101->parent.autonego_ctrl = jl1101_negotiate;
+#else
   jl1101->parent.negotiate = jl1101_negotiate;
+#endif
   jl1101->parent.get_link = jl1101_get_link;
   jl1101->parent.pwrctl = jl1101_pwrctl;
   jl1101->parent.get_addr = jl1101_get_addr;

--- a/esphome/components/ethernet/ethernet_component.cpp
+++ b/esphome/components/ethernet/ethernet_component.cpp
@@ -41,18 +41,27 @@ void EthernetComponent::setup() {
   this->eth_netif_ = esp_netif_new(&cfg);
 
   // Init MAC and PHY configs to default
-  eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
   eth_phy_config_t phy_config = ETH_PHY_DEFAULT_CONFIG();
-
   phy_config.phy_addr = this->phy_addr_;
   phy_config.reset_gpio_num = this->power_pin_;
 
+  eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
+#if ESP_IDF_VERSION_MAJOR >= 5
+  eth_esp32_emac_config_t esp32_emac_config = ETH_ESP32_EMAC_DEFAULT_CONFIG();
+  esp32_emac_config.smi_mdc_gpio_num = this->mdc_pin_;
+  esp32_emac_config.smi_mdio_gpio_num = this->mdio_pin_;
+  esp32_emac_config.clock_config.rmii.clock_mode = this->clk_mode_;
+  esp32_emac_config.clock_config.rmii.clock_gpio = this->clk_gpio_;
+
+  esp_eth_mac_t *mac = esp_eth_mac_new_esp32(&esp32_emac_config, &mac_config);
+#else
   mac_config.smi_mdc_gpio_num = this->mdc_pin_;
   mac_config.smi_mdio_gpio_num = this->mdio_pin_;
   mac_config.clock_config.rmii.clock_mode = this->clk_mode_;
   mac_config.clock_config.rmii.clock_gpio = this->clk_gpio_;
 
   esp_eth_mac_t *mac = esp_eth_mac_new_esp32(&mac_config);
+#endif
 
   switch (this->type_) {
     case ETHERNET_TYPE_LAN8720: {
@@ -76,7 +85,11 @@ void EthernetComponent::setup() {
       break;
     }
     case ETHERNET_TYPE_KSZ8081: {
+#if ESP_IDF_VERSION_MAJOR >= 5
+      this->phy_ = esp_eth_phy_new_ksz80xx(&phy_config);
+#else
       this->phy_ = esp_eth_phy_new_ksz8081(&phy_config);
+#endif
       break;
     }
     default: {
@@ -221,13 +234,13 @@ void EthernetComponent::eth_event_handler(void *arg, esp_event_base_t event_base
       return;
   }
 
-  ESP_LOGV(TAG, "[Ethernet event] %s (num=%d)", event_name, event);
+  ESP_LOGV(TAG, "[Ethernet event] %s (num=%" PRId32 ")", event_name, event);
 }
 
 void EthernetComponent::got_ip_event_handler(void *arg, esp_event_base_t event_base, int32_t event_id,
                                              void *event_data) {
   global_eth_component->connected_ = true;
-  ESP_LOGV(TAG, "[Ethernet event] ETH Got IP (num=%d)", event_id);
+  ESP_LOGV(TAG, "[Ethernet event] ETH Got IP (num=%" PRId32 ")", event_id);
 }
 
 void EthernetComponent::start_connect_() {

